ompi/mca/coll/libnbc/nbc.c

```diff
@@ -534,7 +535,31 @@ static inline int NBC_Start_round(NBC_Handle *handle) {
         } else {
           buf2=opargs.buf2;
         }
-        ompi_op_reduce(opargs.op, buf1, buf2, opargs.count, opargs.datatype);
+
+        /* If the count is > INT_MAX then we need to call ompi_op_reduce()
+         * in iterations of counts <= INT_MAX since it has an `int count`
+         * parameter.
+         */
+        if( OPAL_UNLIKELY(opargs.count > INT_MAX) ) {
+          size_t done_count = 0, shift;
+          int iter_count;
+          ptrdiff_t ext, lb;
+
+          ompi_datatype_get_extent (opargs.datatype, &lb, &ext);
+
+          while(done_count < opargs.count) {
+            if( done_count + INT_MAX > opargs.count ) {
+              iter_count = opargs.count - done_count;
+            } else {
+              iter_count = INT_MAX;
+            }
+            shift = done_count * ext;
+            ompi_op_reduce(opargs.op, buf1 + shift, buf2 + shift, iter_count, opargs.datatype);
```

Comment:
Why not changing the ompi_op_reduce to accept size_t instead ? We will need to change it at some point, starting the 5.0 with this in sounds like the most timely approach.

Reply:
Yeah I can do that for master and v5. Then keep it the same for v4.0 and v4.1 (unless they want to change that interface too). I'll do that in a separate commit so we can cherry-pick whichever the RMs want.

Reply:
:+1: but personally I would not waste my time with anything prior to 5.0.

Reply:
I'm not sure I can make that change and be standard compliant because the `ompi_op_reduce` will pass the `count` parameter to the reduction operation which could be a `MPI_User_function`. I would be passing them up a pointer to a `size_t` instead of an `int`. As long as the value of `count` fits inside an `int` this should be ok, but if it exceeds that then they will see the overflow.

Once we have big count support then we can distinguish between the two cases (`int` vs `MPI_Count`) and do the right thing.

One alternative I thought of was to allow `ompi_op_reduce` to accept a `size_t` and move the looping logic that I put here into that function if `count > INT_MAX`. Then once we get the big count support we can skip the loop if we have a `MPI_Count` reduction function. From my reading of the MPI Standard, it allows us to call the reduction function with a `len < count` to chunk up the data (MPI 4.0, p236, line 44 in section 6.9.5)

Reply:
I just pushed a new commit that implements ^^. Take a look and let me know what you think.

Reply:
Definitively a step in the right direction.